### ISSUE TITLE
Relative time is incorrect in the UI (1.4.0)

### ISSUE THREAD

**Issue by gboucher90** [openzipkin/zipkin#1214]
Hello,

It seems the relative time displayed in the UI is off:
![relative_time_issue](https://cloud.githubusercontent.com/assets/20612880/17335345/e9aba24e-58d9-11e6-8773-569514c34bf1.png)

Though it seems correctly displayed:
![trace_diplay](https://cloud.githubusercontent.com/assets/20612880/17335771/7697fa6c-58db-11e6-8576-1c33155e108d.png)

Here is the raw JSON to reproduce the issue:
[{"traceId":"1e223ff1f80f1c69","name":"post","id":"74280ae0c10d8062","parentId":"bf396325699c84bf","duration":93576,"annotations":[{"endpoint":{"serviceName":"serviceA","ipv4":"x.x.x.x"},"timestamp":1470150004008761,"value":"sr"},{"endpoint":{"serviceName":"serviceA","ipv4":"x.x.x.x"},"timestamp":1470150004102338,"value":"ss"}],"binaryAnnotations":[{"key":"host.name","value":"b","endpoint":{"serviceName":"serviceA","ipv4":"x.x.x.x"}}],"debug":false},{"traceId":"1e223ff1f80f1c69","name":"get","id":"bf396325699c84bf","duration":99410,"annotations":[{"endpoint":{"serviceName":"serviceB","ipv4":"x.x.x.x"},"timestamp":1470150004071068,"value":"sr"},{"endpoint":{"serviceName":"serviceB","ipv4":"x.x.x.x"},"timestamp":1470150004170479,"value":"ss"}],"binaryAnnotations":[{"key":"host.name","value":"a","endpoint":{"serviceName":"serviceB","ipv4":"x.x.x.x"}},{"key":"http.status_code","value":"200","endpoint":{"serviceName":"serviceB","ipv4":"x.x.x.x"}}],"debug":false},{"traceId":"1e223ff1f80f1c69","name":"post","id":"74280ae0c10d8062","parentId":"bf396325699c84bf","duration":94539,"annotations":[{"endpoint":{"serviceName":"serviceA","ipv4":"x.x.x.x"},"timestamp":1470150004074202,"value":"cs"},{"endpoint":{"serviceName":"serviceA","ipv4":"x.x.x.x"},"timestamp":1470150004168741,"value":"cr"}],"binaryAnnotations":[],"debug":false}]

Thank you!

**Comment by codefromthecrypt** [comment]
the service names on your second span's annotations seem incorrect. They all say the same service (servicea). Your client serviceName should be different than your server serviceName, unless this is a loopback.

You'll notice that the timestamp for your "sr" is before your "cs" (which means the server received a request before the client sent it!). I suspect you have clock skew. If you make sure your service names are different between client and server, zipkin will automatically correct the skew.

**Comment by gboucher90** [comment]
Indeed, thank you for your help!

**Comment by gboucher90** [comment]
I tried with a modified json with correct service names, though the relative time is still invalid.

```

[
   {
      "traceId":"1e223ff1f80f1c69",
      "name":"post",
      "id":"74280ae0c10d8062",
      "parentId":"bf396325699c84bf",
      "duration":93576,
      "annotations":[
         {
            "endpoint":{
               "serviceName":"serviceB",
               "ipv4":"192.0.0.0"
            },
            "timestamp":1470150004008761,
            "value":"sr"
         },
         {
            "endpoint":{
               "serviceName":"serviceB",
               "ipv4":"192.0.0.0"
            },
            "timestamp":1470150004102338,
            "value":"ss"
         }
      ],
      "binaryAnnotations":[
      ],
      "debug":false
   }
,
   {
      "traceId":"1e223ff1f80f1c69",
      "name":"get",
      "id":"bf396325699c84bf",
      "duration":99410,
      "annotations":[
         {
            "endpoint":{
               "serviceName":"serviceA",
               "ipv4":"127.0.0.0"
            },
            "timestamp":1470150004071068,
            "value":"sr"
         },
         {
            "endpoint":{
               "serviceName":"serviceA",
               "ipv4":"127.0.0.0"
            },
            "timestamp":1470150004170479,
            "value":"ss"
         }
      ],
      "binaryAnnotations":[
      ],
      "debug":false
   },
   {
      "traceId":"1e223ff1f80f1c69",
      "name":"post",
      "id":"74280ae0c10d8062",
      "parentId":"bf396325699c84bf",
      "duration":94539,
      "annotations":[
         {
            "endpoint":{
               "serviceName":"serviceA",
               "ipv4":"127.0.0.0"
            },
            "timestamp":1470150004074202,
            "value":"cs"
         },
         {
            "endpoint":{
               "serviceName":"serviceA",
               "ipv4":"127.0.0.0"
            },
            "timestamp":1470150004168741,
            "value":"cr"
         }
      ],
      "binaryAnnotations":[
      ],
      "debug":false
   }
]
```

**Comment by gboucher90** [comment]
CorrectForClockSkew.apply() returns a different list depending on the order of  the spans, it can select the client or the server annotations (for the second span). When I submit the input of 3 spans it returns only 2 spans.

**Comment by gboucher90** [comment]
The problem is that if the root span doesn't have any skew which is likely since it starts with only a local span SR/SS then the children don't get adjusted.

ClockSkew skew = getClockSkew(node.value());
returns null for the root span and the processing stops

**Comment by gboucher90** [comment]
Performing the clock skew adjustment even if the current skew is null works for relative time though the duration displayed in the UI is now incorrect.
![duration_ok](https://cloud.githubusercontent.com/assets/20612880/17461781/587298e2-5c98-11e6-88e2-651cfb4f9d9b.png)
![timeline_off](https://cloud.githubusercontent.com/assets/20612880/17461782/58764d34-5c98-11e6-81ee-82f85adfcf0d.png)

**Comment by codefromthecrypt** [comment]
IP addresses have to be different for clock skew to make any changes. It
assumes if you have the same IP address you have the same clock.

**Comment by gboucher90** [comment]
Yes I did use a different set of IP, "127.0.0.0" for serviceA and "192.0.0.0" for serviceB (see json or screenshots)

Anything else I might be missing?

**Comment by codefromthecrypt** [comment]
We don't have any tests to cover a root span which has no client annotations. This could be a bug in the logic.

**Comment by codefromthecrypt** [comment]
basically, I posted the json to a local server and looked in the debugger. It seems the current code only starts thinking when there's skew visible in the root span. In the case of "sr" "ss" spans, you'd not be able to know there's any skew. Hence, I think this might be a bug, especially as we've no test case around this.

We should fix this as "sr" "ss" spans are the most common type of root span!

**Comment by gboucher90** [comment]
Fixed in #1465
